File: XYMenu/XYMenu/XYMenu/XYMenu.m
//
//  XYMenu.m
//  XYMenu
//
//  Created by FireHsia on 2018/1/26.
//  Copyright © 2018年 FireHsia. All rights reserved.
//

#import "XYMenu.h"
#import "XYMenuView.h"

#define kXYMenuScreenWidth [UIScreen mainScreen].bounds.size.width
#define kXYMenuScreenHeight [UIScreen mainScreen].bounds.size.height

static const CGFloat XYMenuWidth = 120; // Menu宽度
static const CGFloat XYMenuItemHeight = 60; // item高度


@interface XYMenu () <UIGestureRecognizerDelegate>

@property (nonatomic, strong) XYMenuView *menuView;
@property (nonatomic, assign) XYMenuType menuType;
@property (nonatomic, assign) CGRect menuInitRect;
@property (nonatomic, assign) CGRect menuResultRect;
@property (nonatomic, assign) BOOL isDismiss;

@end

@implementation XYMenu

// 还是需要一个全局的静态变量

- (instancetype)init
{
    if (self = [super init]) {
        _isDismiss = NO;
        _menuType = XYMenuNormal;
        // 添加pan手势, 防止视图响应scrollview的滚动手势
        UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panAction:)];
        pan.delegate = self;
        [self addGestureRecognizer:pan];
        self.frame = CGRectMake(0, 0, kXYMenuScreenWidth, kXYMenuScreenHeight);
        self.userInteractionEnabled = YES;
        self.backgroundColor = [UIColor clearColor];
        [self addSubview:self.menuView];
    }
    return self;
}

- (void)panAction:(UIPanGestureRecognizer *)pan
{
    if ([pan.view isEqual:self]) {
        return ;
    }
}

- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
{
    if (![self isInMenuViewWithPoint:point]) {
        [XYMenu dismissMenuInView:self.superview];
    }
    return [super hitTest:point withEvent:event];
}

- (BOOL)isInMenuViewWithPoint:(CGPoint)point
{
    NSArray *subViews = self.subviews;
    for (UIView *subView in subViews) {
        if ([subView isKindOfClass:[XYMenuView class]]) {
            XYMenuView *menuView = (XYMenuView *)subView;
            CGPoint menuVPoint = [self convertPoint:point toView:menuView];
            BOOL isInMenu = [menuView pointInside:menuVPoint withEvent:nil];
            return isInMenu;
        }
    }
    return NO;
}

#pragma mark --- 展示菜单
+ (void)showMenuWithImages:(NSArray *)imagesArr titles:(NSArray *)titles inView:(UIView *)view menuType:(XYMenuType)menuType withItemClickIndex:(ItemClickIndexBlock)block
{
    XYMenu *xy_menu = [[XYMenu alloc] init];
    [xy_menu showMenuWithImages:imagesArr titles:titles inView:view menuType:menuType withItemClickIndex:block];
}

+ (void)showMenuWithImages:(NSArray *)imagesArr titles:(NSArray *)titles menuType:(XYMenuType)menuType currentNavVC:(UINavigationController *)currentNavVC withItemClickIndex:(ItemClickIndexBlock)block
{
    
    XYMenu *xy_menu = [[XYMenu alloc] init];
    [xy_menu showMenuWithImages:imagesArr titles:titles menuType:menuType currentNavVC:currentNavVC withItemClickIndex:block];
    
}

#pragma mark --- 隐藏菜单

+ (void)dismissMenuInView:(UIView *)view
{
    XYMenu *menu = [XYMenu XYMenuInView:view];
    if (menu) {
        [menu dimissXYMenu];
    }
}

+ (XYMenu *)XYMenuInView:(UIView *)view
{
    UIView *rootView = [XYMenu rootViewFromSubView:view];
    NSArray *subViews = rootView.subviews;
    for (UIView *view in subViews) {
        if ([view isKindOfClass:[XYMenu class]]) {
            XYMenu *menu = (XYMenu *)view;
            return menu;
        }
    }
    return nil;
}

- (void)dimissXYMenu
{
    if (_isDismiss) return;
    _isDismiss = YES;
    [self.menuView hideContentView];
    self.menuView.alpha = 1.0;
    [UIView animateWithDuration:0.2 animations:^{
        self.menuView.frame = self.menuInitRect;
        self.menuView.alpha = 0.1;
    } completion:^(BOOL finished) {
        _isDismiss = NO;
        [self removeFromSuperview];
    }];
}

- (void)showMenuWithImages:(NSArray *)imagesArr titles:(NSArray *)titles menuType:(XYMenuType)menuType currentNavVC:(UINavigationController *)currentNavVC  withItemClickIndex:(ItemClickIndexBlock)block
{
    CGFloat XYMenuHeight = XYMenuItemHeight * titles.count;
    switch (menuType) {
        case XYMenuLeftNavBar:
        {
            // 左侧NaviBar
            self.menuInitRect = CGRectMake(10 + (XYMenuWidth / 4), 64, 1, 1);
            self.menuResultRect = CGRectMake(10, 64, XYMenuWidth, XYMenuHeight);
        }
            break;
        case XYMenuRightNavBar:
        {
            // 右侧NaviBar
            self.menuInitRect = CGRectMake(kXYMenuScreenWidth - (XYMenuWidth / 4) - 10, 64, 1, 1);
            self.menuResultRect = CGRectMake(kXYMenuScreenWidth - XYMenuWidth - 10, 64, XYMenuWidth, XYMenuHeight);
        }
            break;
        default:
            break;
    }
    [currentNavVC.view addSubview:self];
    __weak typeof(self) weakSelf = self;
    [self.menuView setImagesArr:imagesArr titles:titles withRect:self.menuResultRect withMenuType:menuType withItemClickBlock:^(NSInteger index) {
        [weakSelf dimissXYMenu];
        block(index);
    }];
    
    self.menuView.frame = self.menuInitRect;
    [self.menuView hideContentView];
    self.menuView.alpha = 0.1;
    [UIView animateWithDuration:0.1 animations:^{
        self.menuView.alpha = 1.0;
        self.menuView.frame = self.menuResultRect;
    } completion:^(BOOL finished) {
        [self.menuView showContentView];
    }];
    
}

- (void)showMenuWithImages:(NSArray *)imagesArr titles:(NSArray *)titles inView:(UIView *)view menuType:(XYMenuType)menuType withItemClickIndex:(ItemClickIndexBlock)block
{
    [self configRectWithMenuType:menuType inView:view titles:titles];
    __weak typeof(self) weakSelf = self;
    [self.menuView setImagesArr:imagesArr titles:titles withRect:self.menuResultRect withMenuType:menuType withItemClickBlock:^(NSInteger index) {
        [weakSelf dimissXYMenu];
        block(index);
    }];
    
    self.menuView.frame = self.menuInitRect;
    [self.menuView hideContentView];
    self.menuView.alpha = 0.1;
    [UIView animateWithDuration:0.1 animations:^{
        self.menuView.alpha = 1.0;
        self.menuView.frame = self.menuResultRect;
    } completion:^(BOOL finished) {
        [self.menuView showContentView];
    }];
}

- (void)configRectWithMenuType:(XYMenuType)menuType inView:(UIView *)view titles:(NSArray *)titles
{
    _menuType = menuType;
    UIWindow *keyWindow = [UIApplication sharedApplication].keyWindow;
    UIView *vcView = [XYMenu rootViewFromSubView:view];
    UIView *superView = view.superview;
    CGRect viewRect = view.frame;
    CGRect viewRectFromWindow = [superView convertRect:viewRect toView:keyWindow];
    CGFloat midX = CGRectGetMidX(viewRectFromWindow);
    CGFloat maxX = CGRectGetMaxX(viewRectFromWindow);
    CGFloat maxY = CGRectGetMaxY(viewRectFromWindow);
    CGFloat XYMenuHeight = XYMenuItemHeight * titles.count;
    self.menuInitRect = CGRectMake(maxX, maxY, 1, 1);
    self.menuResultRect = CGRectMake(midX - (XYMenuWidth / 2), maxY + 5, XYMenuWidth, XYMenuHeight);
    [vcView addSubview:self];
}

#pragma mark --- LoadView

- (XYMenuView *)menuView
{
    if (!_menuView) {
        _menuView = [[XYMenuView alloc] initWithFrame:CGRectZero];
    }
    return _menuView;
}

#pragma mark --- 返回当前View的控制器的view
+ (UIView *)rootViewFromSubView:(UIView *)view
{
    UIViewController *vc = nil;
    UIResponder *next = view.nextResponder;
    do {
        if ([next isKindOfClass:[UINavigationController class]]) {
            vc = (UIViewController *)next;
            break;
        }
        next = next.nextResponder;
    } while (next != nil);
    if (vc == nil) {
        do {
            if ([next isKindOfClass:[UIViewController class]]) {
                vc = (UIViewController *)next;
                break;
            }
            next = next.nextResponder;
        } while (next != nil);
    }
    return vc.view;
}

@end

File: XYMenu/XYMenuSwift/XYMenu.swift
//
//  XYMenu.swift
//  XYMenuSwiftDemo
//
//  Created by FireHsia on 2018/2/2.
//  Copyright © 2018年 FireHsia. All rights reserved.
//

import UIKit

enum XYMenuType {
    case XYMenuLeftNavBar
    case XYMenuRightNavBar
    case XYMenuLeftNormal
    case XYMenuMidNormal
    case XYMenuRightNormal
}

private let kXYMenuScreenWidth = Double(UIScreen.main.bounds.size.width)
private let kXYmenuScreenHeight = Double(UIScreen.main.bounds.size.height)
private let XYMenuWidth: Double = 120
private let XYMenuItemHeight: Double = 60

class XYMenu: UIView {
    
    var menuType: XYMenuType
    var menuInitRect: CGRect
    var menuResultRect: CGRect
    var isDismiss: Bool
    var isDown: Bool
    var isShow: Bool
    

    lazy var menuView: XYMenuView = {
        let view = XYMenuView(frame: .zero)
        return view
    }()

    init() {
        isDismiss = false
        isDown = true
        isShow = false
        menuType = .XYMenuMidNormal
        menuInitRect = .zero
        menuResultRect = .zero
        super.init(frame: .zero)
        frame = CGRect(x: 0, y: 0, width: kXYMenuScreenWidth, height: kXYmenuScreenHeight)
        let pan = UIPanGestureRecognizer(target: self, action: #selector(panAction(_:)))
        addGestureRecognizer(pan)
        isUserInteractionEnabled = true
        backgroundColor = .clear
        addSubview(menuView)
    }
    
    @objc func panAction(_ sender : UIPanGestureRecognizer) {
        if (sender.view?.isKind(of: XYMenu.classForCoder()))! {
            return ;
        }
    }
    
    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        if !isInMenuViewWith(point: point) {
            disMissXYMenu()
        }
        return super.hitTest(point, with: event)
    }
    
    fileprivate func isInMenuViewWith(point: CGPoint) -> Bool {
        for subViewV in subviews {
            if (subViewV.isKind(of: XYMenuView.classForCoder())) {
                let menuVPoint = convert(point, to: subViewV)
                let isInMenu = subViewV.point(inside: menuVPoint, with: nil)
                return isInMenu
            }
        }
        return false
    }
    
    fileprivate func disMissXYMenu() {
        if isDismiss || isShow {return}
        isDismiss = true
        menuView.hideContentView()
        menuView.alpha = 1.0
        UIView.animate(withDuration: 0.2, animations: {
            self.menuView.frame = self.menuInitRect
            self.menuView.alpha = 1.0
        }) { (finished) in
            self.isDismiss = false
            self.removeFromSuperview()
        }
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: 展示View菜单
    class func showMenuInView(images: Array<String>, titles: Array<String>, inView: UIView, type: XYMenuType, clickClosure: @escaping ItemClickBlock) {
        let xy_menu = XYMenu.init()
        xy_menu.showMenuInView(images: images, titles: titles, inView: inView, type: type, clickClosure: clickClosure)
    }
    
    // MARK: 展示BarbuttonItem菜单
    class func showMenuInBarButtonItem(images: Array<String>, titles: Array<String>, currentNavVC: UINavigationController, type: XYMenuType, clickClosure: @escaping ItemClickBlock) {
        let xy_menu = XYMenu.init()
        xy_menu.showMenuInBarButtonItem(images: images, titles: titles, currentNavVC: currentNavVC, type: type, clickClosure: clickClosure)
    }
    
    
    fileprivate func showMenuInView(images: Array<String> , titles: Array<String>, inView: UIView, type: XYMenuType, clickClosure: @escaping ItemClickBlock) {
        if isShow {return}
        configRect(type: type, inView: inView, titles: titles)
        showAnimateMenu(images: images, titles: titles, type: type, closure: clickClosure)
    }
    
    fileprivate func showMenuInBarButtonItem(images: Array<String>, titles: Array<String>, currentNavVC: UINavigationController, type: XYMenuType, clickClosure: @escaping ItemClickBlock) {
        if isShow {return}
        let statusRect = UIApplication.shared.statusBarFrame
        let statusHeight = Double(statusRect.size.height)
        let navigationBarHeight = Double(currentNavVC.navigationBar.bounds.size.height)
        let XYMenuHeight = XYMenuItemHeight * Double(titles.count)
        menuType = type
        switch menuType {
        case .XYMenuLeftNavBar:
            menuInitRect = CGRect(x: 10 + (XYMenuWidth / 4), y: statusHeight + navigationBarHeight, width: 1, height: 1)
            menuResultRect = CGRect(x: 10, y: statusHeight + navigationBarHeight, width: XYMenuWidth, height: XYMenuHeight)
        case .XYMenuRightNavBar:
            menuInitRect = CGRect(x: Double(kXYMenuScreenWidth) - 10 - (XYMenuWidth / 4), y: statusHeight + navigationBarHeight, width: 1, height: 1)
            menuResultRect = CGRect(x: Double(kXYMenuScreenWidth) - 10 - XYMenuWidth, y: statusHeight + navigationBarHeight, width: XYMenuWidth, height: XYMenuHeight)
        default:
            break
        }
        currentNavVC.view.addSubview(self)
        showAnimateMenu(images: images, titles: titles, type: type, closure: clickClosure)
    }
    
    fileprivate func configRect(type: XYMenuType, inView: UIView, titles: [String])
    {
        menuType = type
        let vcView = XYMenu.rootViewFromSubView(viewSubView: inView)
        let viewSuperView = inView.superview
        let viewRect = inView.frame
        let viewRectFromWindow = viewSuperView?.convert(viewRect, to: vcView)
        let midX = Double((viewRectFromWindow?.midX)!)
        let minY = Double((viewRectFromWindow?.minY)!)
        let maxY = Double((viewRectFromWindow?.maxY)!)
        let XYMenuHeight = XYMenuItemHeight * Double(titles.count)
        
        if (maxY + XYMenuHeight + 5) > kXYmenuScreenHeight {
            isDown = false
        }
        switch menuType {
        case .XYMenuLeftNormal:
            if isDown {
                menuInitRect = CGRect(x: midX, y: maxY, width: 1, height: 1)
                menuResultRect = CGRect(x: midX - (XYMenuWidth / 4), y: maxY + 5, width: XYMenuWidth, height: XYMenuHeight)
            }else {
                menuInitRect = CGRect(x: midX, y: minY, width: 1, height: 1)
                menuResultRect = CGRect(x: midX - (XYMenuWidth / 4), y: minY - 5 - XYMenuHeight, width: XYMenuWidth, height: XYMenuHeight)
            }
        case .XYMenuMidNormal:
            if isDown {
                menuInitRect = CGRect(x: midX, y: maxY, width: 1, height: 1)
                menuResultRect = CGRect(x: midX - (XYMenuWidth / 2), y: maxY + 5, width: XYMenuWidth, height: XYMenuHeight)
            }else {
                menuInitRect = CGRect(x: midX, y: minY, width: 1, height: 1)
                menuResultRect = CGRect(x: midX - (XYMenuWidth / 2), y: minY - 5 - XYMenuHeight, width: XYMenuWidth, height: XYMenuHeight)
            }
        case .XYMenuRightNormal:
            if isDown {
                menuInitRect = CGRect(x: midX, y: maxY, width: 1, height: 1)
                menuResultRect = CGRect(x: midX - (XYMenuWidth * 3 / 4), y: maxY + 5, width: XYMenuWidth, height: XYMenuHeight)
            }else {
                menuInitRect = CGRect(x: midX, y: minY, width: 1, height: 1)
                menuResultRect = CGRect(x: midX - (XYMenuWidth * 3 / 4), y: minY - 5 - XYMenuHeight, width: XYMenuWidth, height: XYMenuHeight)
            }
        default:
            if isDown {
                menuInitRect = CGRect(x: midX, y: maxY, width: 1, height: 1)
                menuResultRect = CGRect(x: midX - (XYMenuWidth / 2), y: maxY + 5, width: XYMenuWidth, height: XYMenuHeight)
            }else {
                menuInitRect = CGRect(x: midX, y: minY, width: 1, height: 1)
                menuResultRect = CGRect(x: midX - (XYMenuWidth / 2), y: minY - 5 - XYMenuHeight, width: XYMenuWidth, height: XYMenuHeight)
            }
        }
        vcView!.addSubview(self)
    }
    fileprivate func showAnimateMenu(images: [String], titles: [String], type: XYMenuType, closure: @escaping ItemClickBlock) {
        if isShow {return}
        isShow = true
        menuView.setConfig(images: images, titles: titles, rect: menuResultRect, menuType: type, down: isDown) { [unowned self] (index) in
            self.disMissXYMenu()
            closure(index)
        }
        menuView.frame = menuInitRect
        menuView.hideContentView()
        menuView.alpha = 0.1
        UIView.animate(withDuration: 0.2, animations: { [unowned self] in
            self.menuView.frame = self.menuResultRect
            self.menuView.alpha = 1.0
        }) { [unowned self] (finished) in
            self.isShow = false
            self.menuView .showContentView()
        }
    }

    // MARK: 隐藏菜单
    class func disMissMenu(inView: UIView) {
        let xy_menu = XYMenu.XYMenuIn(XYMenuInView: inView)
        if xy_menu != nil {
            xy_menu!.disMissXYMenu()
        }
    }
    
    class func XYMenuIn(XYMenuInView: UIView) -> XYMenu? {
        let rootView = XYMenu.rootViewFromSubView(viewSubView: XYMenuInView)
        if rootView != nil {
            for subV in rootView!.subviews {
                if subV.isKind(of: XYMenu.classForCoder()) {
                    let xy_menu = subV as! XYMenu
                    return xy_menu
                }
            }
        }
        return nil
    }
    
    // MARK: 工具
    class func rootViewFromSubView(viewSubView: UIView) -> UIView? {
        
        
        
        var vc: UIViewController? = nil
        var next = viewSubView.next
        while next != nil {
            if next!.isKind(of: UINavigationController.classForCoder()) {
                vc = next as? UIViewController
                break
            }
            next = next!.next
        }
        if vc == nil {
            next = viewSubView.next
            while next != nil {
                if (next!.isKind(of: UIViewController.classForCoder()) || next!.isKind(of: UITableViewController.classForCoder())) {
                    vc = next as? UIViewController
                    break
                }
                next = next!.next
            }
        }
        if vc != nil {return vc!.view}
        return nil
    }
    
    class func scrollViewFromView(viewSubView: UIView) -> UIScrollView? {
        var scrollView: UIScrollView? = nil
        var next = viewSubView.next
        while next != nil {
            if next!.isKind(of: UIScrollView.classForCoder()) {
                scrollView = next as? UIScrollView
            }
            next = next!.next
        }
        return scrollView
    }
    
}

extension UIBarButtonItem {
    
    func xy_showXYMenu(images: [String], titles: [String], currentNavVC: UINavigationController, type: XYMenuType, closure: @escaping ItemClickBlock) {
        XYMenu.showMenuInBarButtonItem(images: images, titles: titles, currentNavVC: currentNavVC, type: type, clickClosure: closure)
    }
}

extension UIView {
    func xy_showXYMenu(images: [String], titles: [String], type: XYMenuType, closure: @escaping ItemClickBlock) {
        XYMenu.showMenuInView(images: images, titles: titles, inView: self, type: type, clickClosure: closure)
    }
}

